### ISSUE TITLE
A change of `--type-check-only` option's behavior.

### ISSUE THREAD

**Issue by puripuri2100** [gfngfn/SATySFi#170]
SATySFi request the document type with the `-t` option and `--type-check-only` option.

# Bug example

input (file name is `test.saty`)

```
42
```

output

```
$ satysfi -t test.saty
 ---- ---- ---- ----
  target file: 'test.pdf'
  dump file: 'test.satysfi-aux' (will be created)
  parsing 'test.saty' ...
 ---- ---- ---- ----
  type checking 'test.saty' ...
  type check passed. (int)
! [Type Error] file 'test.saty' is not a document file; it is of type
      int
```

---

input (file name is `test2.saty`)

```
@require: stdjareport

document (|
  title = {test};
  author = {puripuri2100};
|) '<>
```

output

```
$ satysfi -t test2.saty
 ---- ---- ---- ----
  target file: 'test2.pdf'
  dump file: 'test2.satysfi-aux' (will be created)
  parsing 'test2.saty' ...
  parsing 'stdjareport.satyh' ...
  parsing 'pervasives.satyh' ...
  parsing 'gr.satyh' ...
  parsing 'geom.satyh' ...
  parsing 'list.satyg' ...
  parsing 'math.satyh' ...
  parsing 'color.satyh' ...
  parsing 'option.satyg' ...
  parsing 'annot.satyh' ...
  parsing 'footnote-scheme.satyh' ...
 ---- ---- ---- ----
  type checking 'list.satyg' ...
  type check passed.
 ---- ---- ---- ----
  type checking 'option.satyg' ...
  type check passed.
 ---- ---- ---- ----
  type checking 'pervasives.satyh' ...
  type check passed.
 ---- ---- ---- ----
  type checking 'color.satyh' ...
  type check passed.
 ---- ---- ---- ----
  type checking 'geom.satyh' ...
  type check passed.
 ---- ---- ---- ----
  type checking 'gr.satyh' ...
  type check passed.
 ---- ---- ---- ----
  type checking 'annot.satyh' ...
  type check passed.
 ---- ---- ---- ----
  type checking 'footnote-scheme.satyh' ...
  type check passed.
 ---- ---- ---- ----
  type checking 'math.satyh' ...
  type check passed.
 ---- ---- ---- ----
  type checking 'stdjareport.satyh' ...
  type check passed.
 ---- ---- ---- ----
  type checking 'test2.saty' ...
  type check passed. (document)
```

# Environment

- ubuntu 18.04.1
- OCaml 4.06.0 + OPAM 2.0.1
- SATySFi commit [`9a7ff4549b97728a9813bdb9ed6bb302e6a14b2a`](https://github.com/gfngfn/SATySFi/commit/9a7ff4549b97728a9813bdb9ed6bb302e6a14b2a)

**Comment by nekketsuuu** [comment]
I think this is not a bug; the type of a top value should be `document` because SATySFi cannot generate a document without that condition. It would be natural to think that if `satysfi --type-check-only test.saty` is passed then `satysfi test.saty` should not be failed during the type check.

Also, if you want to check types of a non-document file, then make its extension to `.satyh` or something :) Then it's no longer a document file and the type check should be passed:

```sh-session
$ cat test.satyh
42
$ satysfi --type-check-only test.satyh
 ---- ---- ---- ----
  target file: 'test.pdf'
  dump file: 'test.satysfi-aux' (already exists)
  parsing 'test.satyh' ...
 ---- ---- ---- ----
  reading 'test.satyh'
  type check passed. (int)
```

**Comment by sakas--** [comment]
This is intended. The purpose of this option is to check whether a *document* is well-typed or/and does not contain a syntax error without generating the document itself. This is useful when you want to implement an on-the-fly syntax checker for an editor since type checking is usually way faster than generating a document.

It wasn't implemented as a function to check the type of a general expression and, in my opinion, this kind of feature should be separated from the `-t` option. Maybe it might be a good idea to implement a REPL for satysfi and implement this kind of feature as a feature of REPL. (Of course, implementing a REPL requires time and effort and if one wants a REPL he/she should open a new issue.)

**Comment by puripuri2100** [comment]
I see.

But this change is not write in `Changelog.md`.
I think I should write this in `Changelog.md`.

**Comment by nekketsuuu** [comment]
@puripuri2100 Is this behavior is changed after `--type-check-only` was introduced?

**Comment by puripuri2100** [comment]
I agree with the correctness of this behavior.
So I modify [SATySFi wiki](https://github.com/gfngfn/SATySFi/wiki/%E3%82%B3%E3%83%9E%E3%83%B3%E3%83%89%E3%83%A9%E3%82%A4%E3%83%B3%E6%9B%B8%E5%BC%8F#t--type-check-only-%E5%9E%8B%E6%A4%9C%E6%9F%BB%E3%81%AE%E3%81%BF%E3%82%92%E8%A1%8C%E3%81%86).

But I want to make a problem is that this behavior has changed.

For older SATySFi (eg [SATySFi for Windows](https://github.com/qnighy/satysfi-cross-windows/releases/tag/20180708)) :

```
$ type test3.saty
42
$ satysfi -t test3.saty
 ---- ---- ---- ----
  target file: 'test3.pdf'
  dump file: 'test3.satysfi-aux' (will be created)
  parsing 'test3.saty' ...
 ---- ---- ---- ----
  reading 'test3.saty' ...
  type check passed. (int)
```

**Comment by sakas--** [comment]
Oh, I didn't notice that the behaviour has changed. Good catch.
It seems that at the time I introduced this option satysfi already reported the ".. is not a document file" error if the top expression does not have document type.  (I checked this with my original branch.)
So, I can't tell when the behaviour has changed. 

Anyway, since the behaviour has changed, I agree that it is worth writing it to `Changelog.md` if we want to make the changelog complete.
